lib/RemoteAST/RemoteAST.cpp

```diff
@@ -1162,6 +1165,35 @@ class RemoteASTContextConcreteImpl final : public RemoteASTContextImpl {
     if (result) return RemoteAddress(*result);
     return getFailure<RemoteAddress>();
   }
+
+  //  OpaqueExistentialContainer layout.
+  //  struct OpaqueExistentialContainer {
+  //  void *fixedSizeBuffer[3];
+  //  Metadata *type;
+  //  WitnessTable *witnessTables[NUM_WITNESS_TABLES];
+  // };
+  Result<std::pair<Type, uint64_t>>
+  getDynamicTypeAndAddressOpaqueExistential(RemoteAddress object) {
+    uint64_t offset = 3 * sizeof(void *);
+    auto MetadataAddressAddress =
+        RemoteAddress(object.getAddressData() + offset);
+    int64_t metadataAddress;
+    if (!readWordOffset(MetadataAddressAddress, &metadataAddress))
+      return getFailure<std::pair<Type, uint64_t>>();
+    auto result = Reader.readTypeFromMetadata(metadataAddress);
+    if (!result)
+      return getFailure<std::pair<Type, uint64_t>>();
+    // XXX: FIXME: offset.
+    return std::make_pair<Type, uint64_t>(std::move(result), 0);
+  }
+
+  Result<std::pair<Type, uint64_t>>
+  getDynamicTypeAndAddressForExistential(RemoteAddress object,
+                                         Type staticType) override {
+    if (staticType->isAnyExistentialType())
+      return getDynamicTypeAndAddressOpaqueExistential(object);
```

Comment:
Okay, the cases here are:
  - It's an existential metatype; the layout is `TargetExistentialMetatypeContainer`.
  - It's an existential value type, i.e. either a `ProtocolType` or a `ProtocolCompositionType`.  You can use the `ExistentialLayout` type in the AST to figure out which of the three layout cases you're in: opaque, class, or error.
    - Opaque uses `TargetOpaqueExistentialContainer`.
    - Class uses `ClassExistentialContainer`, but this apparently hasn't been target-ified (which should be fairly easy).
    - Error is always a `SwiftError*`, which has also not been target-ified.